Code/GraphMol/SynthonSpaceSearch/SynthonSpace.h

```diff
@@ -99,54 +98,60 @@ struct RDKIT_SYNTHONSPACESEARCH_EXPORT SynthonSpaceSearchParams {
                                // search.  0 means no maximum.
 };
 
-// Holds the information about a set of hits.  The molecules can be built
-// by making all combinations of synthons, one taken from each synthon set.
-struct RDKIT_SYNTHONSPACESEARCH_EXPORT SynthonSpaceHitSet {
-  SynthonSpaceHitSet() = delete;
-  SynthonSpaceHitSet(const std::string &id,
-                     const std::vector<std::vector<size_t>> &stu)
-      : reactionId(id), synthonsToUse(stu) {
-    numHits = std::accumulate(
-        synthonsToUse.begin(), synthonsToUse.end(), size_t(1),
-        [](const int prevRes, const std::vector<size_t> &s2) -> size_t {
-          return prevRes * s2.size();
-        });
-  }
-  std::string reactionId;
-  std::vector<std::vector<size_t>> synthonsToUse;
-  size_t numHits{0};
-};
+class Synthon;
 
 class RDKIT_SYNTHONSPACESEARCH_EXPORT SynthonSpace {
  public:
-  // Create the synthonspace from a file in the correct format.
   explicit SynthonSpace() = default;
+  ~SynthonSpace() = default;
   SynthonSpace(const SynthonSpace &other) = delete;
   SynthonSpace &operator=(const SynthonSpace &other) = delete;
-  // Get the number of different reactions in the SynthonSpace.
   /*!
+   * Get the number of different reactions in the SynthonSpace.
    *
    * @return int
    */
-  size_t getNumReactions() const { return d_reactions.size(); }
-  const std::map<std::string, std::unique_ptr<SynthonSet>> &getReactions()
-      const {
-    return d_reactions;
-  }
+  size_t getNumReactions() const;
+  /*!
+   * Get a list of the names of all the reactions in the SynthonSpace.
+   *
+   * @return
+   */
+  std::vector<std::string> getReactionNames() const;
+  const std::shared_ptr<SynthonSet> getReaction(std::string reactionName);
+  // The Synthons have a PatternFingerprint for screening in substructure
+  // searches.  It's important that the screening process creates ones
+  // of the same size, so this finds out what size that is.
+  unsigned int getPatternFPSize() const;
+  // Likewise for the fingerprints used for similarity searching
+  unsigned int getFPSize() const;
 
-  // Get the total number of products that the SynthonSpace could produce.
   /*!
+   * Get the total number of products that the SynthonSpace could produce.
    *
    * @return std::int64_t
    */
-  std::int64_t getNumProducts() const;
+  std::uint64_t getNumProducts() const;
+  /*!
+   * Get the total number of products, formatted nicely in a string.
+   *
+   * @return std::string
+   */
+  std::string getFormattedNumProducts() const;
```

Comment:
seems like a weird one to have in an API
I'd be happier to have this as a function at global scope

Reply:
I find it very useful for this specific purpose.  I guess a function for formatting large ints could be made more widely available, and then it could be used for things like SearchResults.getMaxNumProducts() as well.  Would a function still in the SynthonSpaceSearch namespace be ok, or can you see it being more generally useful?

Reply:
@greglandrum I've moved this but not pushed it yet pending any other review comments.